### ISSUE TITLE
after upgrade to 0.19.0: exceptions.TypeError: string indices must be integers

### ISSUE THREAD

**Issue by kparal** [fedora-infra/fedmsg#463]
When I update from
`fedmsg-0.18.2-1.fc25.noarch`
to
`fedmsg-0.19.0-1.fc25.noarch`
the fedmsg-hub starts failing with this traceback:
```
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/usr/lib64/python2.7/site-packages/twisted/internet/posixbase.py", line 597, in _doReadOrWrite
    why = selectable.doRead()
  File "/usr/lib/python2.7/site-packages/txzmq/connection.py", line 284, in doRead
    log.callWithLogger(self, self.messageReceived, message)
--- <exception caught here> ---
  File "/usr/lib64/python2.7/site-packages/twisted/python/log.py", line 101, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/usr/lib64/python2.7/site-packages/twisted/python/log.py", line 84, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/usr/lib/python2.7/site-packages/txzmq/pubsub.py", line 72, in messageReceived
    self.gotMessage(message[1], message[0])
  File "/usr/lib/python2.7/site-packages/moksha/hub/zeromq/zeromq.py", line 197, in chain_over_moksha_callbacks
    f(_body, _topic)
  File "/usr/lib/python2.7/site-packages/moksha/hub/zeromq/zeromq.py", line 224, in intercept
    return callback(ZMQMessage(_topic, _body))
  File "/usr/lib/python2.7/site-packages/fedmsg/consumers/__init__.py", line 245, in _consume
    self.validate(message)
  File "/usr/lib/python2.7/site-packages/fedmsg/consumers/__init__.py", line 236, in validate
    if not message['topic'] == message['body']['topic']:
exceptions.TypeError: string indices must be integers
```
If I downgrade fedmsg it works OK again.

I tried to debug this and here's `message` printed out from [here](https://github.com/fedora-infra/fedmsg/blob/0.19.0/fedmsg/consumers/__init__.py#L243):
```
<ZMQMessage; topic: u'org.fedoraproject.prod.buildsys.tag', body: u'{"username": "apache", "certificate": "LS...snip...tCg==\\n", "i": 2, "timestamp": 1502280116, "msg_id": "2017-a02ceae4-2ee1-4964-98d0-071ac566987c", "crypto": "x509", "topic": "org.fedoraproject.prod.buildsys.tag", "signature": "EH2Prz1vo3M4HzhBIPAoFhOsqqAkS3TjkH2M2HzcPoJLpaDcHom0BNTeGUEIHp9Vv94otmAtzRSU\\nr8oqXS+hGpZCzu4bYYqzun4SLAzwKeeIm3do2b54uWkNh9XDH9GJDQQg8zNH0lT8/R067EoBgBKy\\n/OV0Hr5vkONp7Ye0wiM=\\n", "msg": {"build_id": 953500, "name": "eom", "tag_id": 427, "instance": "primary", "tag": "f27-pending", "user": "raveit65", "version": "1.19.0", "owner": "raveit65", "release": "4.fc27"}}'>
```

The error occurs at
```
if not message['topic'] == message['body']['topic']
```
because `message['body']` is a string (more precisely `unicode`), not a dict.

**Comment by ralphbean** [comment]
Can you share any more logs that appear before the traceback?

It must be failing to deserialize the body.  There may be some warning before that indicates why (or confirms the suspicion).

**Comment by kparal** [comment]
I don't see anything else useful in the journal:
```
Aug 09 14:04:03 taskotron-dev01.qa.fedoraproject.org systemd[1]: Started Generic fedmsg processing hub.
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]: Unhandled Error
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]: Traceback (most recent call last):
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     return self.currentContext().callWithContext(ctx, func, *args, **kw)
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     return func(*args,**kw)
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib64/python2.7/site-packages/twisted/internet/posixbase.py", line 597, in _doReadOrWrite
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     why = selectable.doRead()
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib/python2.7/site-packages/txzmq/connection.py", line 284, in doRead
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     log.callWithLogger(self, self.messageReceived, message)
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]: --- <exception caught here> ---
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib64/python2.7/site-packages/twisted/python/log.py", line 101, in callWithLogger
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     return callWithContext({"system": lp}, func, *args, **kw)
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib64/python2.7/site-packages/twisted/python/log.py", line 84, in callWithContext
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     return context.call({ILogContext: newCtx}, func, *args, **kw)
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     return self.currentContext().callWithContext(ctx, func, *args, **kw)
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     return func(*args,**kw)
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib/python2.7/site-packages/txzmq/pubsub.py", line 72, in messageReceived
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     self.gotMessage(message[1], message[0])
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib/python2.7/site-packages/moksha/hub/zeromq/zeromq.py", line 197, in chain_over_moksha_callbacks
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     f(_body, _topic)
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib/python2.7/site-packages/moksha/hub/zeromq/zeromq.py", line 224, in intercept
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     return callback(ZMQMessage(_topic, _body))
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib/python2.7/site-packages/fedmsg/consumers/__init__.py", line 245, in _consume
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     self.validate(message)
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:   File "/usr/lib/python2.7/site-packages/fedmsg/consumers/__init__.py", line 236, in validate
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]:     if not message['topic'] == message['body']['topic']:
Aug 09 14:04:31 taskotron-dev01.qa.fedoraproject.org fedmsg-hub[28298]: exceptions.TypeError: string indices must be integers
```

**Comment by jeremycline** [comment]
Thanks for the report! I believe the root cause is https://github.com/fedora-infra/fedmsg/commit/f0553d285e586dc48f1a244b81f1517551d469ee#diff-6d2173d0212e9936a03b9e554a7e2a38L164 which is causing the message body to not be deserialized because it's not bytes. Fix incoming.

**Comment by jeremycline** [comment]
Fixed in #464
